### PR TITLE
SBS add spring security

### PR DESCRIPTION
**Pull request by petro2061**
Adde Spring Security to project. Implemented SecurityConfig class and CustomUserDetailsService that implements UserDetailsService. Updated user model. An implementation of the UserDetails interfase was added to the model and added new fild roles. Created new model Role that represented access to endpoints for roles. Фnd changed book controller - added roles to endpoints.

### CODE REVIEW

File: src/main/java/project/onlinebookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@RequiredArgsConstructor
+@EnableWebSecurity
+@EnableMethodSecurity
+@Configuration
+public class SecurityConfig {
+
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/project/onlinebookstore/model/Role.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/model/Role.java:17]: entity should has get and set methods

File: src/main/java/project/onlinebookstore/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @Column(nullable = false, unique = true)
+    private RoleName role;
+
+    @Override
+    public String getAuthority() {
+        return "ROLE_" + role.name();
+    }
+
+    public enum RoleName {
+        USER,
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/model/Role.java]:
```suggestion
        ROLE_USER,
        ROLE_ADMIN
```

File: src/main/java/project/onlinebookstore/controller/BookController.java [hunk 4]
```diff
@@ -37,6 +38,7 @@ public class BookController {
 
     @Operation(summary = "Get all books",
             description = "Allows get all books using pagination")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/BookController.java]: You can avoid using prefix, Spring Security adds it automatically

File: src/main/java/project/onlinebookstore/security/CustomUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import project.onlinebookstore.repository.user.UserRepository;
+
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
```

File: src/main/java/project/onlinebookstore/security/CustomUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+        return userRepository.findByEmailWithRole(email).orElseThrow(
+                () -> new UsernameNotFoundException("Can't find user with email: "
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
                () -> new EntityNotFoundException("Can't find user with email: "
```

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -23,7 +30,18 @@ public UserResponseDto register(UserRegistrationRequestDto registrationRequestDt
...
         }
-        return userMapper.toUserDto(userRepository
-                .save(userMapper.toUserModel(registrationRequestDto)));
+        Role userRole
+                = roleRepository
+                .findByRole(Role.RoleName.USER)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Can't find role: USER"));
+
+        User savedUserModel = userMapper.toUserModel(registrationRequestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]:
```suggestion
        User user = userMapper.toUserModel(registrationRequestDto);
```

File: src/main/resources/db/changelog/changes/06-insert-admin_user-to-users.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  - changeSet:
+      id: insert-admin_user-to-users
+      author: petro.feniak
+      changes:
+        - insert:
+            tableName: users
+            columns:
+              - column: { name: email, value: "admin@gmail.com" }
+              - column: { name: password,
+                          value: "$2y$10$pCqsETLdAj6d4QvuhPlbDuX4o5HWu98Y8wLs58bRVJ.e2u4IlUXBG"} # "admin_12345!" hashed with BCrypt
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/06-insert-admin_user-to-users.yaml]: remove comment

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

**Comment by petro2061** [comment]
Deleted prefix 'ROLE_' in BookController. Changed naming roles. Added EntityNotFoundException in CustomUserDetailsService.

File: src/main/java/project/onlinebookstore/config/SecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.config.http.SessionCreationPolicy;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@RequiredArgsConstructor
+@EnableWebSecurity
+@EnableMethodSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/config/SecurityConfig.java:17]: Do you need both of this annotations `@EnableWebSecurity and @EnableMethodSecurity`?

File: src/main/java/project/onlinebookstore/security/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import project.onlinebookstore.repository.user.UserRepository;
+
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws EntityNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
```

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -23,7 +30,14 @@ public UserResponseDto register(UserRegistrationRequestDto registrationRequestDt
                     + registrationRequestDto.getEmail()
                     + " already exist");
         }
-        return userMapper.toUserDto(userRepository
-                .save(userMapper.toUserModel(registrationRequestDto)));
+        Role userRole
+                = roleRepository
+                .findByRole(Role.RoleName.ROLE_USER)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Can't find role: USER"));
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]: Use Role.RoleName.ROLE_USER in message

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -23,7 +30,14 @@ public UserResponseDto register(UserRegistrationRequestDto registrationRequestDt
...
+        User user = userMapper.toUserModel(registrationRequestDto);
+        user.setPassword(passwordEncoder.encode(registrationRequestDto.getPassword()));
+        user.getRoles().add(userRole);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]:
```suggestion
        user.setRoles(Set.of(userRole));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve it

**Comment by petro2061** [comment]
Fixed config class and register() method in UserServise class

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -23,7 +31,15 @@ public UserResponseDto register(UserRegistrationRequestDto registrationRequestDt
                     + registrationRequestDto.getEmail()
                     + " already exist");
         }
-        return userMapper.toUserDto(userRepository
-                .save(userMapper.toUserModel(registrationRequestDto)));
+        Role userRole
+                = roleRepository
```
Comment (Sanderfuck) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]:
```suggestion
        Role userRole = roleRepository
```

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Comment by petro2061** [comment]
fixsed problem with codestyle and PreAuthorize annotation

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**